pkg/destroy/aws/aws.go

```diff
@@ -1994,3 +2058,54 @@ func removeSharedTag(ctx context.Context, tagClients []*resourcegroupstaggingapi
 
 	return nil
 }
+
+// destroyedResources stores the results of attempts to destroy resources. The key is the ARN of the resource. The value
+// is whether the resource was successfully destroyed.
+type destroyedResources map[string]bool
+
+// record records where a resource was successfully destroyed. If the specified error is nil, then the resource was
+// destroyed successfully.
+func (r destroyedResources) record(arn string, err error) {
+	r[arn] = (err == nil)
+}
+
+// merge merges the record of destroyed resources from changes into r. If changes and r have results for the same
+// resource, the results from changes have precedence.
+func (r destroyedResources) merge(changes destroyedResources) {
+	for k, v := range changes {
+		r[k] = v
+	}
+}
+
+// extractRemaining removes the resources that could not be successfully deleted from r and returns a slice containing
+// the ARNs of those resources.
+func (r destroyedResources) extractRemaining() []string {
+	if r == nil {
+		return nil
+	}
+	var notDestroyed []string
+	for k, v := range r {
+		if !v {
+			notDestroyed = append(notDestroyed, k)
+			delete(r, k)
```

Comment:
Why the `delete`?  They're still ARNs we have not deleted, so seems like they should stay in `r`.

Reply:
The `delete` is for the case where there was a failure deleting a resource in one iteration of the loop but the resource is no longer present in subsequent iterations of the loop. For example, let's say the destroyer failed to delete resource A in iteration 1. Later in that same iteration, deleting a different resource had the ripple effect of deleting resource A. In iteration 2, there is no attempt to delete resource A since it is not found when looking for tagged resources. When the context expires, the `RunWithContext` function will respond that resource A could not be deleted, even though it no longer exists.

Reply:
> The delete is for the case where there was a failure deleting a resource in one iteration of the loop but the resource is no longer present in subsequent iterations of the loop.

Hmm... I think we may want to iterate over undeleted resources and confirm that they are gone, instead of hoping that we'll discover them again if they aren't gone.  Of course, if we kill and restart the destroyer, we'd loose that state.  But within a given run, I'm concerned about conflating "we don't check it again" with "it is deleted".  Don't hold on this account though; I'll mull it over some more.

Reply:
This retains the same behavior that we have had before. If a resource is not found on a subsequent iteration, then the destroyer makes no further attempt to delete the resource.